File: engine/kirun-engine-js/src/schemas/typeValidator.js
import { SchemaType, path } from "./constants";
import arrayValidator from "./arrayValidator";
import booleanValidator from "./booleanValidator";
import nullValidator from "./nullValidator";
import objectValidator from "./objectValidator";
import stringValidator from "./stringValidator";
import numberValidator from "./numberValidator";

export default function typeValidator(parents, schema, schemaRepository, functionRepository, element, validate) {
  let valid = false;
  let list = [];

  let types = Array.isArray(schema.type) ? schema.type : [schema.type];

  for (let type of types) {
    try {
      typeValidation(parents, type, schema, schemaRepository, functionRepository, element, validate);
      valid = true;
      break;
    } catch (err) {
      valid = false;
      if (Array.isArray(err)) list = [...list, ...err];
      else list.push(err);
    }
  }

  if (!valid) throw [`${path(parents, schema.title)}Value '${element}' is not of valid type(s)`, ...list];
}

function typeValidation(parents, type, schema, schemaRepository, functionRepository, element, validate) {
  if (type === SchemaType.STRING) {
    stringValidator(parents, schema, schemaRepository, functionRepository, element, validate);
  } else if (type === SchemaType.LONG || type === SchemaType.INTEGER || type === SchemaType.DOUBLE || type === SchemaType.FLOAT) {
    numberValidator(type, parents, schema, schemaRepository, functionRepository, element, validate);
  } else if (type === SchemaType.BOOLEAN) {
    booleanValidator(parents, schema, schemaRepository, functionRepository, element, validate);
  } else if (type === SchemaType.OBJECT) {
    objectValidator(parents, schema, schemaRepository, functionRepository, element, validate);
  } else if (type === SchemaType.ARRAY) {
    arrayValidator(parents, schema, schemaRepository, functionRepository, element, validate);
  } else if (type === SchemaType.NULL) {
    nullValidator(parents, schema, element);
  } else {
    throw `${path(parents, schema.title)}${type} is not a valid type.`;
  }
}

File: engine/kirun-engine-java/src/main/java/io/kirun/engine/json/schema/validator/TypeValidator.java
package io.kirun.engine.json.schema.validator;

import static io.kirun.engine.json.schema.validator.SchemaValidator.path;

import java.util.List;

import com.google.gson.JsonElement;

import io.kirun.engine.json.schema.Schema;
import io.kirun.engine.json.schema.type.SchemaType;
import io.kirun.engine.repository.Repository;

public class TypeValidator {

	public static void validate(List<String> parents, SchemaType type, Schema schema, Repository<Schema> repository,
			JsonElement element) {

		if (type == SchemaType.STRING) {
			StringValidator.validate(parents, schema, element);
		} else if (type == SchemaType.LONG || type == SchemaType.INTEGER || type == SchemaType.DOUBLE
				|| type == SchemaType.FLOAT) {
			NumberValidator.validate(type, parents, schema, element);
		} else if (type == SchemaType.BOOLEAN) {
			BooleanValidator.validate(parents, schema, element);
		} else if (type == SchemaType.OBJECT) {
			ObjectValidator.validate(parents, schema, repository, element);
		} else if (type == SchemaType.ARRAY) {
			ArrayValidator.validate(parents, schema, repository, element);
		} else if (type == SchemaType.NULL) {
			NullValidator.validate(parents, schema, element);
		} else {

			throw new SchemaValidationException(path(parents, schema.getTitle()), type + " is not a valid type.");
		}
	}

	private TypeValidator() {
	}
}
